### ISSUE TITLE
Reproduce UDG

### ISSUE THREAD

**Issue by tupbbtus** [Jingkang50/ICCV21_SCOOD#2]
Hi, 
I am trying to replicate the results of the paper with SC-OOD CIFAR-100 benchmarks.
Are the results of papers are not the results of averaged performance over a few runs and are the performance of the best model that shows the best accuracy in the test set? 

Thanks.

**Comment by Jingkang50** [comment]
Thank you for your interest in our work!
The paper report the best results from multiple runs. For each run, we use the best.pth for OOD detection task.

FYI, the work was originally built on the company's own deep learning framework, based on which we report all the results.
We extracted all related code and built this standalone version for release, and checked that most of the results can be reproduced. We noticed that CIFAR-10 can easily match the paper results, but CIFAR-100 benchmark might have a few differences, perhaps due to some minor difference in framework modules and some randomness. We are planning to further explore udg on large-scale datasets based on this codebase and solidify this codebase next month.

**Comment by tupbbtus** [comment]
Hi, thank you for your great work.
I'm still trying to reproduce this work and have 3 questions.

1. The paper says that the purity_idf_thresh value is fixed to 0.8, but the purity_ood_thresh value is not mentioned. Is it okay to use 0.9 as shown in the UDG training config file? Also, in the bash file, purity_ind_thresh and purity_ood_thresh in CIFAR100 are listed as 0.7 and 0.9, respectively. What hyper-parameters were used for the results of the best performance in CIFAR100?

2. The performance of DenseNet is lower than WRN or Res18 in CIFAR100, is this correct? Can I get the UDG performance of DenseNet?

3. OE performance that you report in the paper, is it the best results from multiple runs? or not?

**Comment by Jingkang50** [comment]
Thank you for your questions.

1. We fixed the purity_idf_thresh value to 0.8 without the usage of purity_ood_thresh when we were writing the paper.
We found some performance drop after we extracted all related code from the company's own framework, so we added the purity_ood_thresh to compensate a bit, to enhance the effects of clustering. Based on the released codebase, we tried a few hyperparams and set a generally good value of purity_ind_thresh and purity_ood_thresh as 0.7 and 0.9 for all experiments. 
I am sorry for some mismatches between our original (private) code and the released standalone code (especially on cifar100, we did spend some time to try to figure it out), but the general conclusion in the paper is consistent.

2. The densenet code is borrowed from ODIN codebase, we can just directly run it. We will update our README to report all mean/var for every network comprehensively next month.

3. The paper only reports single value of each method. Honestly, when we did the experiments, we only quickly run the baselines once and focus on our own method tuning. We realize it is not a good case and absolutely should not be advocated, and we will again update the README to make everything clear soon. Thank you for pointing them out!
